### PR TITLE
Cleanup thirdypartyresource before deleting the chart

### PR DESCRIPTION
**Pull request by gianrubio**
When you 

1 - install the chart
2- delete the chart 
3- install the chart again

 you got the error below. This job will delete the tpr before deleting the chart.

```
$ helm install prometheus-operator --name prometheus-operator
Error: jobs.batch "prometheus-operator-prometheus-operator-cleanup-tpr" already exists
```

### CODE REVIEW

**Comment by mgoodness** [comment]
I wrestled with this in a previous iteration of the chart. Do we really want an external process like this to delete TPRs created by the operator? For example, what if there are multiple prometheus-operators running in the cluster? Deleting one would render the other(s) inoperable. Are there any side effects if a user instantiates a `prometheus`, `alertmanager`, or `servicemonitor` TPR, then deletes the TPRs?

Not opposed, mind you. Just want to make sure this wouldn't cause more problems that it would solve.

Note: if we do go ahead, the job's name needs to be randomized to avoid the `already exists` error. Also need to do that for the existing `get-tprs` job.

**Comment by brancz** [comment]
> I wrestled with this in a previous iteration of the chart. Do we really want an external process to delete TPR types created by the operator? For example, what if there are multiple prometheus-operators running in the cluster? Deleting one would render the other(s) inoperable. Are there any side effects if a user instantiates a prometheus, alertmanager, or servicemonitor TPR, then deletes the TPR types?

Note that this is currently not advisable, because the Prometheus Operator does not perform any leader election or similar, however, because of how the Prometheus Operator works, it might still work, but there is no guarantee.

**Comment by gianrubio** [comment]
> For example, what if there are multiple prometheus-operators running in the cluster

When I built this job I didn't realise this scenario, what about an argument to cleanup this `tpr`?

> Note: if we do go ahead, the job's name needs to be randomized to avoid the already exists error. Also need to do that for the existing get-tprs job.

I don't like this idea to randomise names, looks like there's a fix https://github.com/kubernetes/helm/issues/1769#issuecomment-309981012 coming on helm, so we can wait.

**Comment by mgoodness** [comment]
>>For example, what if there are multiple prometheus-operators running in the cluster

>When I built this job I didn't realise this scenario, what about an argument to cleanup this tpr?

Yeah, let's do a bool `cleanupOnDelete`.

>>Note: if we do go ahead, the job's name needs to be randomized to avoid the already exists error. Also need to do that for the existing get-tprs job.

>I don't like this idea to randomise names, looks like there's a fix kubernetes/helm#1769 (comment) coming on helm, so we can wait.

Fair enough. In the meantime, I'll review what you have here so far.

File: helm/prometheus-operator/templates/cleanup.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+kind: Job
+metadata:
+  annotations:
+    helm.sh/hook: pre-delete
+  labels:
+     app: {{ template "name" . }}
+     chart: {{ .Chart.Name }}-{{ .Chart.Version }}
+     heritage: {{ .Release.Service }}
+     release: {{ .Release.Name }}
+  name: {{ template "fullname" . }}-cleanup-tpr
```
Comment (mgoodness) [inline helm/prometheus-operator/templates/cleanup.yaml]: Let's name this (and the pod template below) `{{ template "fullname" . }}-delete-tprs` to be consistent with the existing `{{ template "fullname" . }}-get-tprs` Job.

Also, rename the file to `delete-tprs-job.yaml`.

File: helm/prometheus-operator/templates/cleanup.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        release: {{ .Release.Name }}
+      name: {{ template "fullname" . }}-cleanup-tpr
+    spec:
+      containers:
+      - name: hyperkube
+        image: "{{ .Values.global.hyperkube.repository }}:{{ .Values.global.hyperkube.tag }}"
+        imagePullPolicy: "{{ .Values.global.hyperkube.pullPolicy }}"
+        command:
+          - ./kubectl
+          - delete
```
Comment (mgoodness) [inline helm/prometheus-operator/templates/cleanup.yaml]: Need `- thirdpartyresources` here.

File: helm/prometheus-operator/templates/cleanup.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+          - --ignore-not-found=true
+          - alertmanager.monitoring.coreos.com
+          - prometheus.monitoring.coreos.com
+          - service-monitor.monitoring.coreos.com
+      restartPolicy: OnFailure
+    {{- if .Values.rbacEnable }}
+      serviceAccountName: {{ template "fullname" . }}
+    {{- end }}
```
Comment (mgoodness) [inline helm/prometheus-operator/templates/cleanup.yaml]: Wrap the entire file with a `{{- if .Values.cleanupOnDelete -}}` statement.

**Comment by brancz** [comment]
@mgoodness does this now look good to you?

**Comment by brancz** [comment]
@gianrubio could you update this to delete the CRDs instead of the TPRs?

**Comment by gianrubio** [comment]
Sure, I'll do during the weekend!

**Comment by gianrubio** [comment]
@brancz I did the changes, unfortunatelly I don't have a cluster with `crd` enabled so I can't test the deletion.

**Comment by brancz** [comment]
Thanks! I should be able to test this soon, probably not today, but soon. Again, thanks a lot for taking care of this!

**Pull request closed without merging**